pkg/event/target/mqtt.go

```diff
@@ -179,54 +178,30 @@ func NewMQTTTarget(id string, args MQTTArgs, doneCh <-chan struct{}, loggerOnce
 	// Connect() should be successful atleast once to publish events.
 	token := client.Connect()
 
+	if token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
 	target := &MQTTTarget{
 		id:         event.TargetID{ID: id, Name: "mqtt"},
 		args:       args,
 		client:     client,
 		loggerOnce: loggerOnce,
 	}
 
-	// Retries until the clientID gets registered.
-	retryRegister := func() {
-		// Repeat the pings until the client registers the clientId and receives a token.
-		for {
-			var terr error
-			select {
-			case <-doneCh:
-				return
-			default:
-				terr = token.Error()
-				if token.Wait() && terr == nil {
```

Comment:
The token and token.Error() will be nil always until the wait call is successfully returned. So we could wait for the token to be complete and successful and then see the error if it had any.

```
package main

import (
	"fmt"
	mqtt "github.com/eclipse/paho.mqtt.golang"
)

func main() {
	opts := mqtt.NewClientOptions().AddBroker("tcp://localhost:1883")
	opts.SetClientID("anyclientid")

	c := mqtt.NewClient(opts)
	token := c.Connect()
	// token.Error() will be nil always until the token registration is complete
	fmt.Println("token")
	// The below print will be true always event though the broker connection is off
	fmt.Println(token.Error() == nil)
	// So it is always better to wait for the token registration to complete and then look for errors.
	if token.Wait() && token.Error() != nil {
		fmt.Println(token.Error())
	}

	return
}
```

If we run the above snippet with an inactive broker connection - the `token.Error()` print will always be nil before `token.Wait()` as the token registration process is not yet complete. We could always do `if token.Wait() && token.Error() != nil` to read the error.

Reply:
can i send a PR on this?

Reply:
The flow has changed @Praveenrajmani PTAL